core/common/src/main/java/alluxio/collections/UnmodifiableArrayList.java

```diff
@@ -0,0 +1,311 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.collections;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterators;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import java.util.Collection;
+import java.util.ConcurrentModificationException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.function.Consumer;
+
+/**
+ * This class provides an unmodifiable List proxy for an underlying array.
+ * In other words, this helps you use the array like a List, but not being able to modify it.
+ *
+ * Create an instance with the constructor and get the underlying array with {@link #toArray()}.
+ *
+ * The difference to {@code Collections.unmodifiableList(Arrays.asList(array))} is,
+ * in some Java implementations like adoptjdk, Arrays.asList() copies the array with
+ * {@code new ArrayList<>(array)} so it is not performant.
+ * Also this class makes it possible to recover the original array from the List,
+ * {@code list.toArray()} returns the underlying array instead of a copy.
+ * Use the underlying array for modification if needed.
+ *
+ * Use this when you just want to return an unmodifiable view of the array.
+ *
+ * There is no extra thread safety guarantee on this List.
+ * When the underlying array changes, the view will change too,
+ * and there is no extra thread safety.
+ *
+ * @param <T> element type of the underlying array
+ */
+@NotThreadSafe
+public class UnmodifiableArrayList<T> implements List<T> {
+  T[] mElements;
+
+  /**
+   * @param elements the underlying array
+   */
+  public UnmodifiableArrayList(T[] elements) {
+    Preconditions.checkNotNull(elements);
+    mElements = elements;
+  }
+
+  @Override
+  public int size() {
+    return mElements.length;
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return mElements == null || mElements.length == 0;
+  }
+
+  @Override
+  public boolean contains(Object o) {
+    return indexOf(o) >= 0;
+  }
+
+  @Override
+  public Iterator<T> iterator() {
+    return Iterators.forArray(mElements);
```

Comment:
Can this be used to modify the array?

Reply:
Oops, changed to using the private class `Itr` which disables modification.